# MODULE: Pandas - DATAFRAMES

Pandas (supposedly derived from the words *panel* and *data*) is used for:
- data processing and cleaning, before feeding it into machine learning or data mining algorithms. 
- dealing with missing values
- creating new features (columns) from existing columns
- exploratory data analysis to understand what the data contains, summary statistics and plotting (using Matplotlib).

Data Structures

The `pandas` library contains these useful data structures:
* `Series` object - A 1D array, similar to a column in a spreadsheet (with a column name and row labels).
* `DataFrame` object - A 2D table, similar to a spreadsheet (with column names and row labels).
* `Index` object - 1D array used to store the indices of the data values



**Task for this notebook**

We will learn about DataFrames which contain many columns (each of which is a Series).

We will import 3 data files on US state populations from the web, save it locally and open each up as DataFrames. We will learn techniques to **clean and process the data**, identify and deal with missing values, explore the data and create many new computed variables or measures or features.
These same techniques can be used on any other datasets as well.

**NOTE:**\
Anywhere I write `Series.method()` or `Series.attribute`, it should technically be `pd.Series.method()` and `pd.Series.attribute` since Series and DataFrame are found in the pd package. Similarly `DF.attribute` and `DF.method()` should be `pd.DataFrame.attribute` and `pd.DataFrame.method()`.\
I use the former for simplicity.

# 1. Imports

In [4]:
# imports

import numpy as np
import pandas as pd

In [5]:
pd.__version__

'1.5.3'

# 2. Dataframes (DF)

DataFrames are 2D structures with rows and columns. Columns are returned as `Series` objects.

**Most methods do not modify the object (DF or Series) they are called on.**

Instead, they tend to create and work on a copy and return that copy. This is true of most methods in Pandas. If you want to make changes to the original object, you can set a parameter `inplace` = `True`.

Many methods also allow you to specify which axis you want to work on. `axis = 0` means apply calculation **down rows (within columns)**, and `axis = 1` means apply calculation **across columns (within rows)**.

## 2.1 Obtain some data

We can use CURL to easily download a dataset on US state population from the Internet (at a given remote location / URL).

What is **CURL**?\
CURL stands for Client URL. It  is used in command lines or scripts to transfer data from a URL. When calling curl from within a notebook, type `!curl`.

Syntax:\
`!curl [OPTIONS] [URL}`

options:\
`-O` will save the file in the current working directory with the same file name as remote.\
`-o` lets you specify a different file name or location

Documentation: https://everything.curl.dev/

In [6]:
# You will find these datasets downloaded to your dir

!curl -O https://raw.githubusercontent.com/jakevdp/data-USstates/master/state-population.csv
!curl -O https://raw.githubusercontent.com/jakevdp/data-USstates/master/state-areas.csv
!curl -O https://raw.githubusercontent.com/jakevdp/data-USstates/master/state-abbrevs.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 57935  100 57935    0     0   159k      0 --:--:-- --:--:-- --:--:--  160k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   835  100   835    0     0   3021      0 --:--:-- --:--:-- --:--:--  3047
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   872  100   872    0     0   3263      0 --:--:-- --:--:-- --:--:--  3290


## 2.2 Create Pandas DataFrames

`pd.read_csv()` function is used to read in data into a DataFrame. 

Please check out its documentation for a long list of parameters. Since most of them have default values, we can use them without specifying many arguments other than the file name to read.

In [7]:
pop = pd.read_csv('state-population.csv')
areas = pd.read_csv('state-areas.csv')
abbrevs = pd.read_csv('state-abbrevs.csv')

## 2.3 View the data

View records from the top using `DF.head(n=5)`.\
View records from the end using `DF.tail(n=5)`.\
View randomly sampled records from the end using `DF.sample(n=1)`.\
By default n=5 for the first two, and n=1 for the third, but you can change it.


In [8]:
print(pop.head(), "\n")

print(areas.head(), "\n")

print(abbrevs.tail(n=10), "\n")

# since there is only one parameter, we can even avoid the name 
# of the keyword parameter (n), as long as there is no confusion what parameter
# one is referring to

print(abbrevs.tail(10))  

  state/region     ages  year  population
0           AL  under18  2012   1117489.0
1           AL    total  2012   4817528.0
2           AL  under18  2010   1130966.0
3           AL    total  2010   4785570.0
4           AL  under18  2011   1125763.0 

        state  area (sq. mi)
0     Alabama          52423
1      Alaska         656425
2     Arizona         114006
3    Arkansas          53182
4  California         163707 

            state abbreviation
41   South Dakota           SD
42      Tennessee           TN
43          Texas           TX
44           Utah           UT
45        Vermont           VT
46       Virginia           VA
47     Washington           WA
48  West Virginia           WV
49      Wisconsin           WI
50        Wyoming           WY 

            state abbreviation
41   South Dakota           SD
42      Tennessee           TN
43          Texas           TX
44           Utah           UT
45        Vermont           VT
46       Virginia           VA
47     Was

You will notice that all the DFs have as index 0,1,2,3..... That is because we did not specify to use any column as the index, so it assigned a default index.

Another function available in Jupyter is `display()`.
It is a built-in function available in Jupyter to pretty-display DataFrames.

In [9]:
display(pop)

state/region     ages  year   population
0              AL  under18  2012    1117489.0
1              AL    total  2012    4817528.0
2              AL  under18  2010    1130966.0
3              AL    total  2010    4785570.0
4              AL  under18  2011    1125763.0
...           ...      ...   ...          ...
2539          USA    total  2010  309326295.0
2540          USA  under18  2011   73902222.0
2541          USA    total  2011  311582564.0
2542          USA  under18  2012   73708179.0
2543          USA    total  2012  313873685.0

[2544 rows x 4 columns]

## 2.4 Explore the data

### 2.4.1 counts
`DF.count()` - will give us the count of nonp-missing values in each column of the DF.

We can see below that areas DF and abbrevs DF have a different number of states. One has 51 and another has 52.
pop DF has multiple columns for each state/region-ages-year combination.

In [10]:
pop.count()

state/region    2544
ages            2544
year            2544
population      2524
dtype: int64

In [11]:
areas.count()

state            52
area (sq. mi)    52
dtype: int64

In [12]:
abbrevs.count()

state           51
abbreviation    51
dtype: int64

### 2.4.2 categorical counts

For a given categorical column called 'colname', we can obtain unique counts (like a histogram) of each value using `DF['colname'].value_counts()`.

#### accessing a single column

In [13]:
# we have seen this before - using indexing
pop['ages']

# can also use the . operator (columns are attributes of the DF)
pop.ages    

# Note: if we don't use tail() or head() it obtains and prints all values - trauncated

0       under18
1         total
2       under18
3         total
4       under18
         ...   
2539      total
2540    under18
2541      total
2542    under18
2543      total
Name: ages, Length: 2544, dtype: object

In [14]:
pop['ages'].value_counts()

under18    1272
total      1272
Name: ages, dtype: int64

### 2.4.3 unique values

`DF.nunique()` gives the number of unique values in each column of the DF.\
`Series.unique()` gives an array of the unique values in one selected column of the DF.\

In [15]:
pop.nunique()

state/region      53
ages               2
year              24
population      2524
dtype: int64

In [16]:
# number of unique values in just one column - state/region

pop['ages'].nunique()

2

In [17]:
pop['state/region'].unique()

array(['AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'DC', 'FL', 'GA',
       'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 'LA', 'ME', 'MD', 'MA',
       'MI', 'MN', 'MS', 'MO', 'MT', 'NE', 'NV', 'NH', 'NJ', 'NM', 'NY',
       'NC', 'ND', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX',
       'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY', 'PR', 'USA'],
      dtype=object)

NOTE that PR, DC and USA are included, in addition to 50 states to make up 53 values.\
missing values or `"na"` are not included in count of unique.

In [18]:
print(abbrevs['abbreviation'].nunique())
abbrevs['abbreviation'].unique()

51


array(['AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'DC', 'FL', 'GA',
       'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 'LA', 'ME', 'MT', 'NE',
       'NV', 'NH', 'NJ', 'NM', 'NY', 'NC', 'ND', 'OH', 'OK', 'OR', 'MD',
       'MA', 'MI', 'MN', 'MS', 'MO', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX',
       'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY'], dtype=object)

In [19]:
print(areas['state'].nunique())
areas['state'].unique()

52


array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina',
       'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming',
       'District of Columbia', 'Puerto Rico'], dtype=object)

Observe that these three DF contain a different number of unique entries for state or state abbreviation. 

- pop DF has 3 extra values for state/region: USA, PR, DC than areas DF,
- areas DF in turn contains one more state than abbrevs DF: District of Columbia

We will need to keep these in mind when combining these DF.

## 2.5 Combining / Merging DataFrames

Now we have three separate DataFrames, and they contain some columns in common (such as state, state/region), and have some different columns (population, area etc) that we may wish to combine into a single DF.



There are two ways to do this - using `DF.join()` and `DF.merge()`.

The `join()` method works best when we are joining dataframes on their indexes (though you can specify another column to join on for the left dataframe; the right dataframe is joined using its index).

The `merge()` method is more versatile and allows us to specify columns besides the index to join on for both dataframes.

### 2.5.1 pd.merge()

`pd.merge()` function accepts as parameters the names of the two DF to merge.

We can specify the columns to use to merge. If the column names are different in the DFs, we can use `left_on` and `right_on` as parameters to give the names of columns to merge on.\
If the colnames are the same, we can just use parameter `on`.

parameter `how` specifies how to merge - it can take on values: 
* **left**: use only keys from left frame, similar to a SQL left outer join; preserve key order.
* **right**: use only keys from right frame, similar to a SQL right outer join; preserve key order.
* **outer**: use union of keys from both frames, similar to a SQL full outer join; sort keys lexicographically.
* **inner**: use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys.
* **cross**: creates the cartesian product from both frames, preserves the order of the left keys.


Let us first merge the **pop** and **abbrevs** dataframes.

We will specify `how='outer'` to retain also rows that dont match in both DF - so as to keep all values


In [20]:
popabbrevs = pd.merge(pop, abbrevs, how='outer',
                      left_on='state/region', 
                      right_on='abbreviation')
#OR
popabbrevs = pop.merge(abbrevs, how='outer',
                      left_on='state/region', 
                       right_on='abbreviation')
popabbrevs.head()

state/region     ages  year  population    state abbreviation
0           AL  under18  2012   1117489.0  Alabama           AL
1           AL    total  2012   4817528.0  Alabama           AL
2           AL  under18  2010   1130966.0  Alabama           AL
3           AL    total  2010   4785570.0  Alabama           AL
4           AL  under18  2011   1125763.0  Alabama           AL

We will perform  another merge to add the area of each state or region. We can match both DFs - popabbrevs and areas - on the common column labeled state. So, this time, we can use the `on` parameter instead of `left_on` and `right_on`.

Again, we use `how='outer'` merge to retain key values present in one or both DFs


In [21]:
popabbrevsareas = pd.merge(popabbrevs, areas, how='outer', on='state')
popabbrevsareas.head()

state/region     ages    year  population    state abbreviation  \
0           AL  under18  2012.0   1117489.0  Alabama           AL   
1           AL    total  2012.0   4817528.0  Alabama           AL   
2           AL  under18  2010.0   1130966.0  Alabama           AL   
3           AL    total  2010.0   4785570.0  Alabama           AL   
4           AL  under18  2011.0   1125763.0  Alabama           AL   

   area (sq. mi)  
0        52423.0  
1        52423.0  
2        52423.0  
3        52423.0  
4        52423.0

NOTE: The datatype of year has changed to float - indicates that **null or missing values** were added to the column 'year'.
Pandas doesnt allow an int column to contain missing ints.

### 2.5.2 make a copy of a DF

Now that we have merged the 3 DataFrames, let us create a copy of the new merged DataFrame so as to preserve the merged original data.

Create a copy of the data using `DF.copy()`

In [22]:
popDF = popabbrevsareas.copy()

### 2.5.3 examine some properties/ attributes of the merged DF

#### shape

Just as with NumPy arrays, `DF.shape` gives us the dimensions or axes of the DF

In [23]:
popDF.shape

(2545, 7)

#### dtypes

`DF.dtypes` gives us the datatypes of each column in the DF, returned as a Series

In [24]:
popDF.dtypes

state/region      object
ages              object
year             float64
population       float64
state             object
abbreviation      object
area (sq. mi)    float64
dtype: object

#### columns

`DF.columns` will list all columns. 

It will provide one column as an `Index` object. Remember the Index object is used to row and column indices or labels.

In [25]:
popDF.columns

Index(['state/region', 'ages', 'year', 'population', 'state', 'abbreviation',
       'area (sq. mi)'],
      dtype='object')

#### index

In [26]:
popDF.index

Int64Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
            ...
            2535, 2536, 2537, 2538, 2539, 2540, 2541, 2542, 2543, 2544],
           dtype='int64', length=2545)

## 2.6 Indexing to select rows and/or columns

DF has two axes - (rows, columns)

Just as with Series, 
- use `DF.loc` when selecting rows/columns using their label index 
- and `DF.iloc` when selecting rows/columns using their integer index.

When selecting all rows for some columns, use `DF[colname(s)]` or `DF.loc[: , colname(s)]` \
or `DF.iloc[ : , colindex(es)]`.

You can specify multiple colname(s) and colindex(es) with either slicing or fancy indexing.
 

In [27]:
# integer slicing for rows and integer fancy indexing for columns
popDF.iloc[0:2 ,[1, 3]]

ages  population
0  under18   1117489.0
1    total   4817528.0

In [28]:
# selecting using column label
popDF['abbreviation'].head()

0    AL
1    AL
2    AL
3    AL
4    AL
Name: abbreviation, dtype: object

In [29]:
# can use .loc when selecting multiple column labels, dont forget to provide an index for row and col
popDF.loc[:, ['abbreviation', 'area (sq. mi)']].head()
# same as 
popDF[['abbreviation', 'area (sq. mi)']].head()

abbreviation  area (sq. mi)
0           AL        52423.0
1           AL        52423.0
2           AL        52423.0
3           AL        52423.0
4           AL        52423.0

Cannot mix integer and label indexing - produces an error

In [30]:
#popDF[0:2, ['abbreviation', 'area (sq. mi)']]

Slicing indexing applied to columns - requires `DF.loc[: , colstart : colend : colstep]`

In [31]:
# correct command is below
popDF.loc[ :, 'state/region' : 'population'].tail()   # this will work to select the slice of labels from columns

state/region     ages    year   population
2540          USA  under18  2011.0   73902222.0
2541          USA    total  2011.0  311582564.0
2542          USA  under18  2012.0   73708179.0
2543          USA    total  2012.0  313873685.0
2544          NaN      NaN     NaN          NaN

In [32]:
# this will not work, as now it is looking for the labels in rows - does not exist
#popDF['state/region' : 'population'].tail()    

Can mix different types of indexing as long both are applied to labels or integers.

In [33]:
# For example boolean indexing for rows and fancy label indexing for columns
# requires DF.loc[]

print(popDF.loc[popDF['state/region'].isin(['USA','DC']) , ['abbreviation', 'state/region']], "\n")

     abbreviation state/region
384            DC           DC
385            DC           DC
386            DC           DC
387            DC           DC
388            DC           DC
...           ...          ...
2539          NaN          USA
2540          NaN          USA
2541          NaN          USA
2542          NaN          USA
2543          NaN          USA

[96 rows x 2 columns] 



## 2.7 Dropping rows/ columns

In [34]:
popDF.count()  

state/region     2544
ages             2544
year             2544
population       2524
state            2449
abbreviation     2448
area (sq. mi)    2449
dtype: int64

Observe that there are missing values in 'abbreviation' but not in 'state/region' - since they are almost duplicates, let us drop 'abbreviations', and rename 'state/region' as 'abbrev', which is what 'state/region' column really is.

Use `DF.drop()` to specify columns or rows to drop. Use the following parameters
 - `axis` to specify whether to search for labels in columns (=1) or rows (=0). 
 - `inplace` = True to make the change in the DataFrame itself, without creating a copy with the change.
 - `errors` set to ignore to not raise an error if not found, otherwise it will.


### 2.7.1 drop columns

In [35]:
popDF.drop('abbreviation', axis=1, inplace=True, errors ='ignore')
popDF.head()

state/region     ages    year  population    state  area (sq. mi)
0           AL  under18  2012.0   1117489.0  Alabama        52423.0
1           AL    total  2012.0   4817528.0  Alabama        52423.0
2           AL  under18  2010.0   1130966.0  Alabama        52423.0
3           AL    total  2010.0   4785570.0  Alabama        52423.0
4           AL  under18  2011.0   1125763.0  Alabama        52423.0

#### rename columns

To rename columns, use `DF.rename()` which accepts a dict or mapper between old names and new names.\
Parameters:
 - parameter `axis` = 1 to look for old names in columns. 
 - parameter `inplace` will make the change in the DataFrame itself.

In [36]:
popDF.rename({'state/region' : 'abbrev'}, axis = 1, inplace = True)

In [37]:
popDF.tail()

abbrev     ages    year   population        state  area (sq. mi)
2540    USA  under18  2011.0   73902222.0          NaN            NaN
2541    USA    total  2011.0  311582564.0          NaN            NaN
2542    USA  under18  2012.0   73708179.0          NaN            NaN
2543    USA    total  2012.0  313873685.0          NaN            NaN
2544    NaN      NaN     NaN          NaN  Puerto Rico         3515.0

### 2.7.2 drop rows

Let us drop the rows with column 'abbrev' = 'USA' since this is not a state, and we only want to keep 50 states and DC and PR.

In popDF, rows dont have labels, we only have an int index. So, in order to drop rows we need to give `DF.drop()` the **index of the row(s)** that we want dropped. 

We dont know the exact labels/indices but we can find them using `popDF[condition].index`.\
specify the boolean condition as `popDF['abbrev']=='USA'`.

WARNING: when rows are dropped the indexes are not updated (since they are treated as labels), so the labels will no longer be continuous.

In [38]:
popDF[popDF['abbrev']=='USA'].index

Int64Index([2496, 2497, 2498, 2499, 2500, 2501, 2502, 2503, 2504, 2505, 2506,
            2507, 2508, 2509, 2510, 2511, 2512, 2513, 2514, 2515, 2516, 2517,
            2518, 2519, 2520, 2521, 2522, 2523, 2524, 2525, 2526, 2527, 2528,
            2529, 2530, 2531, 2532, 2533, 2534, 2535, 2536, 2537, 2538, 2539,
            2540, 2541, 2542, 2543],
           dtype='int64')

In [39]:
index_to_drop = popDF.loc[popDF['abbrev'] =='USA'].index
f"There are {len(index_to_drop)} values to drop"

popDF.drop(index_to_drop, axis=0, inplace=True, errors ='ignore')

# or in one line, write the below:
#popDF.drop(popDF[popDF['abbrev'] =='USA'].index, axis=0, inplace=True, errors ='ignore')

In [40]:
# let's make sure rows with USA in 'abbrev' column are dropped. They were at the end, so call .tail()
display(popDF.tail())
print(popDF.count(), "\n")
print(popDF.shape)

abbrev     ages    year  population        state  area (sq. mi)
2492     PR  under18  2011.0    869327.0          NaN            NaN
2493     PR    total  2011.0   3686580.0          NaN            NaN
2494     PR  under18  2012.0    841740.0          NaN            NaN
2495     PR    total  2012.0   3651545.0          NaN            NaN
2544    NaN      NaN     NaN         NaN  Puerto Rico         3515.0

abbrev           2496
ages             2496
year             2496
population       2476
state            2449
area (sq. mi)    2449
dtype: int64 

(2497, 6)


### 2.7.2 drop empty rows

Also drop any rows that have missing values in all columns. we can do this using `DF.dropna()`.with parameters
 - `how` which can be set to 'any' or 'all'. 
 - `axis = 0` to look for missing values in rows (or 1 for columns)

In [41]:
popDF.dropna(how = "all", axis=0).shape

(2497, 6)

## 2.8 Sorting

We can sort a DataFrame by either its index or the values in its columns.


### 2.8.1 sort by index

To sort on a DF's index, use `DF.sort_index()` with parameter
 - `ascending` = True. The default order is ascending, and it can be changed to `False`.

In [42]:
popDF.sort_index(ascending = False).head()

abbrev     ages    year  population        state  area (sq. mi)
2544    NaN      NaN     NaN         NaN  Puerto Rico         3515.0
2495     PR    total  2012.0   3651545.0          NaN            NaN
2494     PR  under18  2012.0    841740.0          NaN            NaN
2493     PR    total  2011.0   3686580.0          NaN            NaN
2492     PR  under18  2011.0    869327.0          NaN            NaN

### 2.8.2 sort by columns

To sort on a DF's columns, use `DF.sort_values()` with parameterss
 - `by` = colname(s), You can specify more than one column as a list.
 - `ascending` = [True, ...]  one boolean value for each column listed in `by`
   

In [43]:
popDF.sort_values(by=['year', 'ages'], ascending = [True, False]).head()

abbrev     ages    year  population       state  area (sq. mi)
46      AL  under18  1990.0   1050041.0     Alabama        52423.0
49      AK  under18  1990.0    177502.0      Alaska       656425.0
142     AZ  under18  1990.0   1006040.0     Arizona       114006.0
144     AR  under18  1990.0    620933.0    Arkansas        53182.0
238     CA  under18  1990.0   7980501.0  California       163707.0

## 2.9 Dealing with missing values

Are there any columns that contain missing values?

We can use `DF.isna()` or `DF.notna()`.
This will return a boolean DF of the same size as the original DF. 

We can then combine this with `DF.any()` or `DF.all()` with parameter
 - `axis` to specify 0 (row/index; default) or 1 (columns) - which axis should be reduced

any() will return True if any of the values are True (non-zero or non-empty)\
all() will return True if all of the values are True

In [44]:
# any(axis = 0) will check within columns whether there are any missing values
# all columns appear to have some missing values

popDF.isna().any(axis=0)

abbrev           True
ages             True
year             True
population       True
state            True
area (sq. mi)    True
dtype: bool

Are there any rows that contain all missing values?

In [45]:
# no rows contain all NAs
# here we use chaining to produce an answer

print(popDF.isna().all(axis=1))

# to find out how many are False, call Series.sum()
popDF.isna().all(axis=1).sum()

aDF = popDF.isna()
bSeries = aDF.all(axis=1)
cint = bSeries.sum()

0       False
1       False
2       False
3       False
4       False
        ...  
2492    False
2493    False
2494    False
2495    False
2544    False
Length: 2497, dtype: bool


0

Identify rows that contain more than 3 missing values in the columns

In [46]:
popDF.loc[popDF.isna().sum(axis = 1) > 3]

abbrev ages  year  population        state  area (sq. mi)
2544    NaN  NaN   NaN         NaN  Puerto Rico         3515.0

The other missing values are to rows with 'abbrev' = PR

In [47]:
display(popDF.loc[popDF['abbrev'] == 'PR'])

abbrev     ages    year  population state  area (sq. mi)
2448     PR  under18  1990.0         NaN   NaN            NaN
2449     PR    total  1990.0         NaN   NaN            NaN
2450     PR    total  1991.0         NaN   NaN            NaN
2451     PR  under18  1991.0         NaN   NaN            NaN
2452     PR    total  1993.0         NaN   NaN            NaN
2453     PR  under18  1993.0         NaN   NaN            NaN
2454     PR  under18  1992.0         NaN   NaN            NaN
2455     PR    total  1992.0         NaN   NaN            NaN
2456     PR  under18  1994.0         NaN   NaN            NaN
2457     PR    total  1994.0         NaN   NaN            NaN
2458     PR    total  1995.0         NaN   NaN            NaN
2459     PR  under18  1995.0         NaN   NaN            NaN
2460     PR  under18  1996.0         NaN   NaN            NaN
2461     PR    total  1996.0         NaN   NaN            NaN
2462     PR  under18  1998.0         NaN   NaN            NaN
2463     PR    total  1998.0         NaN   NaN            NaN
2464     PR    total  1997.0         NaN   NaN            NaN
2465     PR  under18  1997.0         NaN   NaN            NaN
2466     PR    total  1999.0         NaN   NaN            NaN
2467     PR  under18  1999.0         NaN   NaN            NaN
2468     PR    total  2000.0   3810605.0   NaN            NaN
2469     PR  under18  2000.0   1089063.0   NaN            NaN
2470     PR    total  2001.0   3818774.0   NaN            NaN
2471     PR  under18  2001.0   1077566.0   NaN            NaN
2472     PR    total  2002.0   3823701.0   NaN            NaN
2473     PR  under18  2002.0   1065051.0   NaN            NaN
2474     PR    total  2004.0   3826878.0   NaN            NaN
2475     PR  under18  2004.0   1035919.0   NaN            NaN
2476     PR    total  2003.0   3826095.0   NaN            NaN
2477     PR  under18  2003.0   1050615.0   NaN            NaN
2478     PR    total  2005.0   3821362.0   NaN            NaN
2479     PR  under18  2005.0   1019447.0   NaN            NaN
2480     PR    total  2006.0   3805214.0   NaN            NaN
2481     PR  under18  2006.0    998543.0   NaN            NaN
2482     PR    total  2007.0   3782995.0   NaN            NaN
2483     PR  under18  2007.0    973613.0   NaN            NaN
2484     PR    total  2008.0   3760866.0   NaN            NaN
2485     PR  under18  2008.0    945705.0   NaN            NaN
2486     PR  under18  2013.0    814068.0   NaN            NaN
2487     PR    total  2013.0   3615086.0   NaN            NaN
2488     PR    total  2009.0   3740410.0   NaN            NaN
2489     PR  under18  2009.0    920794.0   NaN            NaN
2490     PR    total  2010.0   3721208.0   NaN            NaN
2491     PR  under18  2010.0    896945.0   NaN            NaN
2492     PR  under18  2011.0    869327.0   NaN            NaN
2493     PR    total  2011.0   3686580.0   NaN            NaN
2494     PR  under18  2012.0    841740.0   NaN            NaN
2495     PR    total  2012.0   3651545.0   NaN            NaN

Both state and area are missing.
- We know what the state and area is for PR (from one of the rows in our merged data), so we can fill it in.

### 2.9.1 set missing values to new values

In [48]:
# let us replace the missing values for the state column for column 'abbrev' = PR

popDF.loc[popDF['abbrev'] == 'PR', 'state'] = 'Puerto Rico'

### 2.9.2 use a fill-value to replace missing values

Fill in missing values using `DF.fillna()`

We can fill missing values by specifying a common fill value for all missing values in any columns. Parameter 
 - `inplace` = False by default, so original DataFrame is not change

In [49]:
# fill any missing value (in any column) to 0 in a copy of the DF
popDF.fillna(0, inplace = False)

# fill missing values in column 'area' only in a copy of the DF
popDF['area (sq. mi)'].fillna(0)

# different fill-values for different columns by specifying a dict of colname: value to fill, in a copy of the DF
popDF.fillna({'population':-1, 'area (sq. mi)':0}, inplace = False )


abbrev     ages    year  population        state  area (sq. mi)
0        AL  under18  2012.0   1117489.0      Alabama        52423.0
1        AL    total  2012.0   4817528.0      Alabama        52423.0
2        AL  under18  2010.0   1130966.0      Alabama        52423.0
3        AL    total  2010.0   4785570.0      Alabama        52423.0
4        AL  under18  2011.0   1125763.0      Alabama        52423.0
...     ...      ...     ...         ...          ...            ...
2492     PR  under18  2011.0    869327.0  Puerto Rico            0.0
2493     PR    total  2011.0   3686580.0  Puerto Rico            0.0
2494     PR  under18  2012.0    841740.0  Puerto Rico            0.0
2495     PR    total  2012.0   3651545.0  Puerto Rico            0.0
2544    NaN      NaN     NaN        -1.0  Puerto Rico         3515.0

[2497 rows x 6 columns]

### 2.9.3 using backfill technique with `DF.groupby()` and `DF.fillna()` 

#### DF.groupby()

Similar to the SQL language, Pandas allows grouping your data into groups to run calculations over each group using the `DF.groupby()` method. This returns to us a **DataFrameGroupBy**, which is a different class than **DataFrame**, and has its own methods. 

A much more sophisticated approach for filling is shown here:
we can  use ***back filling*** or ***forward filling*** by setting parameter `method` of `DF.fillna()` to 'bfill' or 'ffill'

However we only want to use data within groups (states) to backfill area. First, we need to use `DF.groupby('state')` to group our data by 'state'. Then select just the column(s) that you want to fill - here, select the area column. Then call `DF.fillna()` and set the `method` parameter to 'bfill'.

This will fill values that are missing with the next valid value encountered within the column, and within the same group. Finally, save the results back to the area column in the original DF.


In [50]:
type(popDF.groupby('state')[['area (sq. mi)']])

pandas.core.groupby.generic.DataFrameGroupBy

In [51]:
popDF['area (sq. mi)'] = popDF.groupby('state')['area (sq. mi)'].fillna(method='bfill')

# let's verify
popDF.loc[popDF.state == 'Puerto Rico'].tail(10)

abbrev     ages    year  population        state  area (sq. mi)
2487     PR    total  2013.0   3615086.0  Puerto Rico         3515.0
2488     PR    total  2009.0   3740410.0  Puerto Rico         3515.0
2489     PR  under18  2009.0    920794.0  Puerto Rico         3515.0
2490     PR    total  2010.0   3721208.0  Puerto Rico         3515.0
2491     PR  under18  2010.0    896945.0  Puerto Rico         3515.0
2492     PR  under18  2011.0    869327.0  Puerto Rico         3515.0
2493     PR    total  2011.0   3686580.0  Puerto Rico         3515.0
2494     PR  under18  2012.0    841740.0  Puerto Rico         3515.0
2495     PR    total  2012.0   3651545.0  Puerto Rico         3515.0
2544    NaN      NaN     NaN         NaN  Puerto Rico         3515.0

### Drop unwanted rows

Let us drop the last row with index - 2544, since we no longer need it.

[Side note: Say we cannot obtain PR population data prior to 2000 - so we may wish to drop those rows too.]

Recall that `DF.drop()` requires the index or the label to drop. It can drop either rows (axis = 0) or columns (axis = 1). We want to drop from rows (axis=0) where the value of of 'abbrev' and 'year' are NA

In [52]:
popDF.dropna (how ='any', axis = 0, inplace = True)
print(popDF.count())

abbrev           2476
ages             2476
year             2476
population       2476
state            2476
area (sq. mi)    2476
dtype: int64


In [53]:
popDF.dtypes

abbrev            object
ages              object
year             float64
population       float64
state             object
area (sq. mi)    float64
dtype: object

In [54]:
# now that the year col has no missing values, we can change its type back to int
popDF['year'] = popDF['year'].astype(int)
popDF['population'] = popDF['population'].astype(int)

## 2.10 Summary and aggregate statistics 

### 2.10.1 Basic info on columns

`DF.info()` provides a summary of the data.

In [55]:
print(popDF.info(),'\n')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2476 entries, 0 to 2495
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   abbrev         2476 non-null   object 
 1   ages           2476 non-null   object 
 2   year           2476 non-null   int32  
 3   population     2476 non-null   int32  
 4   state          2476 non-null   object 
 5   area (sq. mi)  2476 non-null   float64
dtypes: float64(1), int32(2), object(3)
memory usage: 116.1+ KB
None 



`DF.describe()` provides distribution details for values in numeric columns in DF. It leaves out non-numeric columns

In [56]:
popDF.describe()

year    population  area (sq. mi)
count  2476.000000  2.476000e+03    2476.000000
mean   2001.556543  3.482132e+06   73452.686591
std       6.917905  4.986552e+06   94687.159589
min    1990.000000  1.013090e+05      68.000000
25%    1996.000000  7.306692e+05   35387.000000
50%    2002.000000  1.557804e+06   56276.000000
75%    2008.000000  4.373440e+06   84904.000000
max    2013.000000  3.833252e+07  656425.000000

### 2.10.2 DF.agg()

`DF.agg()` is used to calculate multiple **aggregate (reduction)** statistics, and the desired stats are given as a list argument. Keep in mind you can only call reduction stats with `DF.agg()` 

Depending on the statistic, it might ignore non-numeric columns (for mean, var, std). Or it might raise warnings.

In [57]:
#popDF.agg(['max','min', 'mean'])

You can use `DF.select_dtypes()` to select columns of certain datatypes only - to avoid the above warning.

In [58]:
popDF_numeric = popDF.select_dtypes([int, float])
popDF_numeric.sample(3)

year  population  area (sq. mi)
1225  2002     1424513        69709.0
2397  2011     5708785        65503.0
1444  2010     2062013         8722.0

In [59]:
# now call DF.add on this new DF
popDF_numeric.agg(['min', 'max', 'sum', 'mean', 'std'])

year    population  area (sq. mi)
min   1.990000e+03  1.013090e+05   6.800000e+01
max   2.013000e+03  3.833252e+07   6.564250e+05
sum   4.955854e+06  8.621760e+09   1.818689e+08
mean  2.001557e+03  3.482132e+06   7.345269e+04
std   6.917905e+00  4.986552e+06   9.468716e+04

In [60]:
# or you can also select the columns that you want using fancy indexing
popDF[['year', 'population', 'area (sq. mi)']].agg(['min', 'max'])

year  population  area (sq. mi)
min  1990      101309           68.0
max  2013    38332521       656425.0

### 2.10.3 DF.groupby() with DF.agg()

We can also first group by a subset of columns, and then call `DF.groupby().agg()`.

When you call .groupby(“key”) on your DataFrame, you’re not creating a new DataFrame. What you get is a **DataFrameGroupBy** object which is sort of a view of your DataFrame that “splits” it into groups by “key”. No actual computation happens at this point, and to get a result, you need to specify what operation should be applied to this object for the actual computations to happen and to get a combined result. You can use any aggregate function here, like .sum() or mean(), and you can also use such GroupBy methods as aggregate, filter, transform and apply.


In [61]:
popDF.groupby(popDF.year)['population'].mean().head()

year
1990    3.076876e+06
1991    3.120529e+06
1992    3.166896e+06
1993    3.210917e+06
1994    3.252615e+06
Name: population, dtype: float64

In [62]:
popDF.groupby(popDF.year)['population'].agg(['min', 'max']).head()

min       max
year                  
1990  112632  29959515
1991  116825  30470736
1992  118636  30974659
1993  120471  31274928
1994  122170  31484435

## 2.11 Re-shaping DataFrame

### 2.11.1 Wide format using pivot tables

The data we have would be called **long** format since there are multiple rows where observations have repeated values for a subset of the columns.
We can reshape into **wide** format, using `DF.pivot()`.

`DF.pivot()` reshapes data (produces a "pivot" table) based on column values. It has a few useful parameters:
 - `index` values (can be a list) are used to form the row axes of the resulting DataFrame.
 - `columns` values (can be a list) are used to form the column axes of the resulting DataFrame.
 - `values` specifies which columns from original DF we want to keep or include (this is value we want in our pivot table cells)
  
Multiple values in either the row or columns axes in the pivot table will result in a ***MultiIndex*** in that axis.

Let us compute the population and area (values) separately for each age group - total and under18, for each State/year combination.

In [63]:
popDF_wide_mi = popDF.pivot(index = ['abbrev', 'year'], columns = 'ages', values = ['population', 'area (sq. mi)' ])
popDF_wide_mi.head()

population           area (sq. mi)          
ages             total   under18         total   under18
abbrev year                                             
AK     1990   553290.0  177502.0      656425.0  656425.0
       1991   570193.0  182180.0      656425.0  656425.0
       1992   588736.0  184878.0      656425.0  656425.0
       1993   599434.0  187190.0      656425.0  656425.0
       1994   603308.0  187439.0      656425.0  656425.0

### 2.11.2 Multi-index

Notice that this creates a new DataFrame with a **multi-index** in the rows (where the row index is now formed using multiple values - 'abbrev' and 'year').\
Similarly, column appears to also have a multi-level label. Level 0 column label is ['population', 'area(sq.mi)'], whereas Level 1  column labels is ['total', 'under18', 'total', 'under18'].


#### removing multilevel index and columns

we can get rid of multi-level labels in columns by dropping a level from columns using `DF.columns.droplevel()`

In [64]:
popDF_wide_mi.columns = popDF_wide_mi.columns.droplevel(0)
popDF_wide_mi.head()

ages            total   under18     total   under18
abbrev year                                        
AK     1990  553290.0  177502.0  656425.0  656425.0
       1991  570193.0  182180.0  656425.0  656425.0
       1992  588736.0  184878.0  656425.0  656425.0
       1993  599434.0  187190.0  656425.0  656425.0
       1994  603308.0  187439.0  656425.0  656425.0

In [65]:
popDF_wide_mi.columns

Index(['total', 'under18', 'total', 'under18'], dtype='object', name='ages')

In [66]:
# we can just create new column names if we like.

popDF_wide_mi.columns = ['totalpop', 'under18pop', 'area', 'area2']

In [67]:
# since the area is the same, irrespective of which population we use, 
# we can retain just one of the population columns and drop the other

popDF_wide_mi.drop('area2', axis=1, inplace= True, errors = 'ignore')
popDF_wide_mi.head()

totalpop  under18pop      area
abbrev year                                
AK     1990  553290.0    177502.0  656425.0
       1991  570193.0    182180.0  656425.0
       1992  588736.0    184878.0  656425.0
       1993  599434.0    187190.0  656425.0
       1994  603308.0    187439.0  656425.0

### 2.11.3 Changing the index

####  setting index 

If the index gets reset to a counter (0,1,2,3....) post some operation, we can use `DF.set_index(colnametouseasindex)` to change it to a specific column that we want that has unique values. 

#### reset index

Above, we have what we call a multi-index, formed by a combination of values across multiple columns.  Multi-indices are a bit tricky to work with. Luckily, you can convert a DF with multi-index back to a regular DataFrame with a single index, by calling `DF.reset_index()` - which will give you indices 0,1,2,.....
The columns that formed the multi-index become regular columns.

In [68]:
popDF_wide = popDF_wide_mi.reset_index()
popDF_wide.head()

abbrev  year  totalpop  under18pop      area
0     AK  1990  553290.0    177502.0  656425.0
1     AK  1991  570193.0    182180.0  656425.0
2     AK  1992  588736.0    184878.0  656425.0
3     AK  1993  599434.0    187190.0  656425.0
4     AK  1994  603308.0    187439.0  656425.0

We will continue to work with this (wide) DF in the tasks below.

## 2.12 Adding new calculated columns

Pandas gives us several ways including DF methods that we can use to create new features or calculated columns. Here are a few:
 - use NumPy ufuncs directly with shorthand notations for their operations when available
 
 
 - `DF.transform()` -  used to call function on self producing a DF with transformed values and that has the same axis length as self. DF.transform() cannot be applied with aggregate functions (such as mean, min, max, var, std, etc) as transform will not return a smaller size DF. transform operates on a single column (a Series) at a time, you cannot, for instance, subtract a value of one column from another via transform. 
 
 When combined with `DF.groupby().transform()` produces interesting results.  
 
 
 - `DF.assign()` -  This function will always create a new DataFrame and will not modify the original. It allows you to create and add multiple new columns at the same time. It allows the subsequent new column to be created using the new column created before it in the DF.assign() line of code. It allows method chaining based on newly created columns. It passes through the whole DataFrame to the calculation, so we can specify row or column-based operations, which is very convenient
 
 
 - `DF.apply()` - The custom function passed to apply can return a scalar, or a Series or DataFrame (or numpy array or even list). This is very flexible as we get to pass it the entire DF so we can perform any kind of operation on either axis, even involving multiple columns or rows. DF.apply() can returned a result of a different shape, e.g. aggregated. However, this is a very slow operation, and not considered efficient. SO when ufuncs or Pandas builtin funcs exist, you should prefer those.
 
 
 

### 2.12.1 Using NumPy ufuncs

When you want to add columns to an existing DF, one by one, you can write:\
`DF[newcol1] = DF-operation`\
`DF[newcol2] = DF-operation`\
or \
`DF[[newcol1, newcol2]] = DF-operation that produces 2 columns`

where DF-operation is one or more NumPy built-in ufuncs for arithmetic, logical, mathematical operations etc. NumPy (and SciPy) contain many mathematical and statistical functions that are useful. Recall that Series or DF can be passed into NumPy functions as an argument.

This ends up modifying the original DF as we are adding new columns to it using =.

In [69]:
# sending Pandas columns (as Series or DF) as argument to a NumPy function
# the outpout is a new DF with 2 columns
# notice how it uses the same original column names

np.log(popDF_wide[['totalpop', 'under18pop']]).head()

totalpop  under18pop
0  13.223638   12.086737
1  13.253730   12.112750
2  13.285733   12.127451
3  13.303741   12.139879
4  13.310183   12.141209

In [70]:
np.divide(popDF_wide['totalpop'], popDF_wide['area'])
# same as 
popDF_wide['density'] = popDF_wide['totalpop'] / popDF_wide['area']
popDF_wide.head()

abbrev  year  totalpop  under18pop      area   density
0     AK  1990  553290.0    177502.0  656425.0  0.842884
1     AK  1991  570193.0    182180.0  656425.0  0.868634
2     AK  1992  588736.0    184878.0  656425.0  0.896882
3     AK  1993  599434.0    187190.0  656425.0  0.913180
4     AK  1994  603308.0    187439.0  656425.0  0.919081

In [71]:
popDF_wide['log_density'] = np.log(popDF_wide['totalpop'] / popDF_wide['area'])
popDF_wide.head()

abbrev  year  totalpop  under18pop      area   density  log_density
0     AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926
1     AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834
2     AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831
3     AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823
4     AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381

In [72]:
np.subtract(popDF_wide['totalpop'], popDF_wide['under18pop'])
# same as 
popDF_wide['gte18pop'] = popDF_wide['totalpop'] - popDF_wide['under18pop']

popDF_wide.head()

abbrev  year  totalpop  under18pop      area   density  log_density  \
0     AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926   
1     AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834   
2     AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831   
3     AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823   
4     AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381   

   gte18pop  
0  375788.0  
1  388013.0  
2  403858.0  
3  412244.0  
4  415869.0

### 2.12.2 DF.groupby()

Sometimes, we will want to create new columns that are the result of grouping operations where we group by one or more columns. For this we can use `DF.groupby()`, which returns a **DataFrameGroupBy** object, which is a different class than a DataFrame.

Grouping operations typically include some combination of:
 - *split* the object, 
 - *apply* a function or method, and 
 - *combine* the results. 
 
You can apply methods such as: `GroupByDF.mean()`, `GroupByDF.min()`, `GroupByDF.max()`, `GroupByDF.var()`, `GroupByDF.std()` etc. groupby() combines the results from each grouping for us. These methods are found in the DataFrameGroupBy class.

Let us produce the averages within each year (over all states and territories) for each of the columns

In [73]:
popDF_wide.groupby('year').mean().head()

C:\Users\u0625859\AppData\Local\Temp\ipykernel_37872\2888749276.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  popDF_wide.groupby('year').mean().head()


totalpop    under18pop          area     density  log_density  \
year                                                                      
1990  4.894565e+06  1.259187e+06  74252.627451  313.547813     4.287649   
1991  4.960411e+06  1.280647e+06  74252.627451  313.596818     4.300957   
1992  5.029691e+06  1.304102e+06  74252.627451  313.982921     4.315244   
1993  5.096443e+06  1.325391e+06  74252.627451  314.720642     4.329437   
1994  5.159330e+06  1.345901e+06  74252.627451  314.278067     4.342872   

          gte18pop  
year                
1990  3.635378e+06  
1991  3.679763e+06  
1992  3.725589e+06  
1993  3.771052e+06  
1994  3.813429e+06

Produce the averages within each year (over all states and territories) for only the 3 population columns

In [74]:
# If we only want the mean of the 3 population columns, we can first select them.
# then group by year
# then call a function that we want to apply within groups.

# the below line of code will not work - why?
# popDF_wide[['totalpop', 'under18pop', 'gte18pop']].groupby('year').mean().head()

# because when a column is not included in the DF, you cannot refer to it just by name later in the code
# We did not include 'year' in the columns selected.

# correct code - both are equivalent
popDF_wide[['totalpop', 'under18pop', 'gte18pop']].groupby(popDF_wide['year']).mean().head()
#or
popDF_wide[['totalpop', 'under18pop', 'gte18pop', 'year']].groupby('year').mean().head()

totalpop    under18pop      gte18pop
year                                          
1990  4.894565e+06  1.259187e+06  3.635378e+06
1991  4.960411e+06  1.280647e+06  3.679763e+06
1992  5.029691e+06  1.304102e+06  3.725589e+06
1993  5.096443e+06  1.325391e+06  3.771052e+06
1994  5.159330e+06  1.345901e+06  3.813429e+06

### 2.12.3 DF.transform()

Let's see how DF.transform() works.

`DF.transform()` will produce a DataFrame that will have same axis length as self (that is it performs a 1:1 operation). It accepts as parameter 
 - `func`: the type of transformation that you want to perform. This can be a custom or lambda function, string built-in Pandas function name, or  a list of multiple function names ([np.exp, 'sqrt'] etc . DF.transform operates on one Series column at a time.
 - `axis` :  default = 0, i.e within each column. 
 
 Since it produces a new Series/DF of the same length as original Series/DF, the output can be added as a new column(s) in the DataFrame.

In [75]:
# transform can accept a single function as argument or multiple functions as a list
popDF_wide['totalpop'].transform('log')

0       13.223638
1       13.253730
2       13.285733
3       13.303741
4       13.310183
          ...    
1233    13.235426
1234    13.243203
1235    13.248695
1236    13.264949
1237    13.275356
Name: totalpop, Length: 1238, dtype: float64

In [76]:
# transform can accept a single function as argument or multiple functions as a list
# they can be given as numpy ufuncs, Pandas funcs as a string or custom or lambda funcs

popDF_wide['totalpop'].transform([np.log, 'sqrt', lambda x : x ** (1/2)])

log        sqrt    <lambda>
0     13.223638  743.834659  743.834659
1     13.253730  755.111250  755.111250
2     13.285733  767.291340  767.291340
3     13.303741  774.231232  774.231232
4     13.310183  776.729039  776.729039
...         ...         ...         ...
1233  13.235426  748.231916  748.231916
1234  13.243203  751.147123  751.147123
1235  13.248695  753.212453  753.212453
1236  13.264949  759.358940  759.358940
1237  13.275356  763.320378  763.320378

[1238 rows x 3 columns]

#### Combining DF.groupby() with DF.transform() - a unique behavior of DF. transform()

Note that when we call DF.groupby(), and then call a reduction/aggregation function, it produces a new DataFrame of grouped values of a smaller size than self (the input DataFrame). Why? because reduction/aggregation is M:1 operation. So we cannot add a column of grouped means back to our original DataFrame because they are of different sizes.

What if we want to add a new column of group means to the original DataFrame without reducing its size? For that we can use `DF.transform()` after `DF.groupby()`

You cannot specify a reduction function (M:1 operation) inside a `DF.transform()`. You have to give it functions that can perform 1:1 operations.

However, when combined with groupby as  `DF.groupby().transform()` it produces interesting results.  Here transform() accepts a reduction function, and then broadcasts the reduced value for each group to the entire group. It finally returns a Series of DF of the same length /size as the input DF after combining the output of all the groups.


In [77]:
# adding 'mean' as a function in Series.transform() produces an error - try it for yourself by uncommenting below
#popDF_wide['totalpop'].transform('mean')

In [78]:
# here, we are adding a new column of the mean of the total population for each year, by group.
# notice how the mean per group is broadcasted to the whole group
popDF_wide['avgstatepopbyyear'] = popDF_wide.groupby('year')['totalpop'].transform('mean')
popDF_wide.head()

abbrev  year  totalpop  under18pop      area   density  log_density  \
0     AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926   
1     AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834   
2     AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831   
3     AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823   
4     AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381   

   gte18pop  avgstatepopbyyear  
0  375788.0       4.894565e+06  
1  388013.0       4.960411e+06  
2  403858.0       5.029691e+06  
3  412244.0       5.096443e+06  
4  415869.0       5.159330e+06

In [79]:
# verifying that the avgstatepopbyyear is the same for all states within a year
popDF_wide[popDF_wide.year == 1990].head()

abbrev  year    totalpop  under18pop      area     density  log_density  \
0      AK  1990    553290.0    177502.0  656425.0    0.842884    -0.170926   
24     AL  1990   4050055.0   1050041.0   52423.0   77.257215     4.347140   
48     AR  1990   2356586.0    620933.0   53182.0   44.311722     3.791249   
72     AZ  1990   3684097.0   1006040.0  114006.0   32.314940     3.475530   
96     CA  1990  29959515.0   7980501.0  163707.0  183.006927     5.209524   

      gte18pop  avgstatepopbyyear  
0     375788.0       4.894565e+06  
24   3000014.0       4.894565e+06  
48   1735653.0       4.894565e+06  
72   2678057.0       4.894565e+06  
96  21979014.0       4.894565e+06

### 2.12.4 DF.assign()

There is another way to create new columns using `DF.assign()`. This function will always create a new DataFrame and will not modify the original. \
Benefits:
- It allows you to create and add multiple new columns at the same time.
- It allows the subsequent new column to be created using the new column created before it in the DF.assign() line of code
- It allows method chaining based on newly created columns.
- It passes through the whole DataFrame to the calculation, so we can specify row or column-based operations, which is very convenient. 

The syntax is as follows:\
`DF.assign( colname1_without_quotes = lambda function,\
            colname2_without_quotes = lambda function,\
            ...)`
            
Recall the syntax of the **lambda function**:\
`lambda x : f(x)`\
where x is the input to the lambda, and f(x) is the computation we want to perform using x.

The input to `DF.assign()` is always the whole DataFrame DF. That means in the code below x is the original DataFrame, popDF_wide.

So let's see how to combine DF.assign with lambda fuctions.

In [80]:
# column names cannot start with a number, hence I added _ to the second column created below.

popDF_wide = popDF_wide.assign(under18pct = lambda x : x['under18pop']/x['totalpop'], 
                               gte18pct = lambda x : x['gte18pop']/x['totalpop'])
popDF_wide.head()

abbrev  year  totalpop  under18pop      area   density  log_density  \
0     AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926   
1     AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834   
2     AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831   
3     AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823   
4     AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381   

   gte18pop  avgstatepopbyyear  under18pct  gte18pct  
0  375788.0       4.894565e+06    0.320812  0.679188  
1  388013.0       4.960411e+06    0.319506  0.680494  
2  403858.0       5.029691e+06    0.314025  0.685975  
3  412244.0       5.096443e+06    0.312278  0.687722  
4  415869.0       5.159330e+06    0.310685  0.689315

### 2.12.5 DF.apply()
This is a very flexible way to apply any kind of function to any axis in the DF - rows or columns. 

When to use Df.apply()?\
We can use `DF.apply()` to apply a custom function (or lambda function) along an axis of the DataFrame, i.e. when no easy built-in functions exist. If built-in vectorized operations exist, then you should prefer those. For example, we would not want to compute standard deviation by hand, since there is np.std() available.

While apply is a very flexible method, its downside is that using it can be quite a bit slower than using more specific methods like agg or transform. Pandas offers a wide range of methods that will be much faster than using apply for their specific purposes, so try to use them before reaching for apply.

DF.apply() passes the whole Dataframe to the calculation, and applies the function to all columns. `axis = 0` by default.

An example:

Say, for each state we want to compute the ratio of maxpop/minpop for three columns - Total, under18 and 18andover populations, over all the years.

First select the columns that we want from the original DF.

The calculation to apply in `DF.apply()` can  be specified as a lambda function. Recall that these are simple anonymous functions that return one value (can take any number of inputs/arguments). The input to lambda here is the GroupDF of 3 columns. The lambda function will run the calculation on each of the 3 columns for each GroupDataFrame.


In [81]:
popDF_wide[['abbrev', 'totalpop','under18pop', 'gte18pop']]\
        .groupby('abbrev')\
        .apply(lambda df: df.max()/df.min())

totalpop  under18pop  gte18pop
abbrev                                
AK      1.328656    1.085261  1.455608
AL      1.193495    1.080841  1.240741
AR      1.255788    1.146576  1.296058
AZ      1.798711    1.618873  1.870688
CA      1.279477    1.180189  1.326613
CO      1.592798    1.404124  1.661365
CT      1.092380    1.130800  1.117549
DC      1.143688    1.220227  1.193586
DE      1.382608    1.245037  1.433092
FL      1.500222    1.348731  1.545740
GA      1.534282    1.425510  1.574414
HI      1.260948    1.111542  1.315870
IA      1.111254    1.020667  1.147810
ID      1.592416    1.368851  1.694330
IL      1.124752    1.103408  1.158162
IN      1.182285    1.121266  1.209748
KS      1.166284    1.098225  1.193080
KY      1.189832    1.082169  1.230412
LA      1.095685    1.158821  1.164801
MA      1.112047    1.111701  1.142364
MD      1.235229    1.171992  1.266609
ME      1.080205    1.192494  1.155224
MI      1.079902    1.156295  1.116575
MN      1.234751    1.097377  1.288839
MO      1.178458    1.087299  1.218764
MS      1.159878    1.057250  1.221401
MT      1.268633    1.070388  1.372328
NC      1.477797    1.405831  1.501020
ND      1.137852    1.169119  1.208356
NE      1.181364    1.079708  1.219328
NH      1.192382    1.153514  1.269709
NJ      1.146384    1.170975  1.156851
NM      1.370480    1.143814  1.477241
NV      2.285695    2.105020  2.353817
NY      1.090470    1.105519  1.121697
OH      1.065044    1.093419  1.103307
OK      1.222859    1.125116  1.258519
OR      1.373968    1.166692  1.450683
PA      1.073131    1.083441  1.104790
PR      1.058586    1.337803  1.037744
RI      1.068175    1.162173  1.078659
SC      1.363790    1.172562  1.432123
SD      1.211986    1.060908  1.279856
TN      1.327202    1.225283  1.362004
TX      1.550599    1.435318  1.597148
UT      1.675620    1.429688  1.815308
VA      1.328705    1.226127  1.361920
VT      1.109476    1.234627  1.195555
WA      1.421853    1.226077  1.492604
WI      1.170892    1.051863  1.231348
WV      1.035777    1.144412  1.086225
WY      1.284265    1.110983  1.401014

### 2.12.6 DF.groupby() + DF.apply() + DF.assign()  (Read if interested)

**this is the trickiest calculation in this file**

Let's add a new column that is the year over year percentage change in the gte18pop for each state.

This is somewhat tricky as it requires a few things.
- There is a Series method called `Series.pct_change()` which computes the % change between the current and previous value in a column.
    - great, that will work, but our data is grouped by states, and when the rows change states, we wouldn't want to compute % change across two states. That would be meaningless.
        - So, we can prevent that by specifying that our data is grouped across rows - using `DF.groupby()`
        - So, we can prevent that by specifying that our data is grouped across rows - using `DF.groupby()`
        
        
- Next, we want to create and add this new column to the original DF. For this we need to use `DF.assign()`, but it is ***not possible to call DF.assign on DataFrameGroupBy***. \
However, we can use `GroupByDF.apply()` to apply a custom or lambda function to the DF. The lambda function accepts a DataFrame and can call `.assign()` to add a new column to itself

Since `DF.assign()` creates a new DF, we can save the result back in popDF_wide (the original DF).


In [82]:
popDF_wide = popDF_wide.groupby('abbrev')\
                        .apply(lambda df: df.assign(gte18_pctgr = df['gte18pop'].pct_change()))
popDF_wide.head(100)


C:\Users\u0625859\AppData\Local\Temp\ipykernel_37872\4120072540.py:1: FutureWarning: Not prepending group keys to the result index of transform-like apply. In the future, the group keys will be included in the index, regardless of whether the applied function returns a like-indexed object.
To preserve the previous behavior, use

	>>> .groupby(..., group_keys=False)

To adopt the future behavior and silence this warning, use 

	>>> .groupby(..., group_keys=True)
  popDF_wide = popDF_wide.groupby('abbrev')\


abbrev  year    totalpop  under18pop      area     density  log_density  \
0      AK  1990    553290.0    177502.0  656425.0    0.842884    -0.170926   
1      AK  1991    570193.0    182180.0  656425.0    0.868634    -0.140834   
2      AK  1992    588736.0    184878.0  656425.0    0.896882    -0.108831   
3      AK  1993    599434.0    187190.0  656425.0    0.913180    -0.090823   
4      AK  1994    603308.0    187439.0  656425.0    0.919081    -0.084381   
..    ...   ...         ...         ...       ...         ...          ...   
95     AZ  2013   6626624.0   1616814.0  114006.0   58.125221     4.062600   
96     CA  1990  29959515.0   7980501.0  163707.0  183.006927     5.209524   
97     CA  1991  30470736.0   8245605.0  163707.0  186.129707     5.226444   
98     CA  1992  30974659.0   8439647.0  163707.0  189.207908     5.242846   
99     CA  1993  31274928.0   8624810.0  163707.0  191.042093     5.252494   

      gte18pop  avgstatepopbyyear  under18pct  gte18pct  gte18_pctgr  
0     375788.0       4.894565e+06    0.320812  0.679188          NaN  
1     388013.0       4.960411e+06    0.319506  0.680494     0.032532  
2     403858.0       5.029691e+06    0.314025  0.685975     0.040836  
3     412244.0       5.096443e+06    0.312278  0.687722     0.020765  
4     415869.0       5.159330e+06    0.310685  0.689315     0.008793  
..         ...                ...         ...       ...          ...  
95   5009810.0       6.148922e+06    0.243988  0.756012     0.015365  
96  21979014.0       4.894565e+06    0.266376  0.733624          NaN  
97  22225131.0       4.960411e+06    0.270607  0.729393     0.011198  
98  22535012.0       5.029691e+06    0.272469  0.727531     0.013943  
99  22650118.0       5.096443e+06    0.275774  0.724226     0.005108  

[100 rows x 12 columns]

## 2.13 Binning (Bin values into discrete intervals)

Sometimes, we may want to discretize a continuous variable. We can do this by using a couple of top-level Pandas functions:
- `pd.cut()` takes a 1D array of Series, and produces equal size bins from the values in the Series
- `pd.qcut()`takes a 1D array of Series, and produces approx equal number of observations in each bin (quantile-based)

Both functions accept a 1D array and require us to provide the number of bins.

In [83]:
pd.cut(popDF_wide['under18pct'], 5).value_counts().sort_index()

(0.167, 0.206]     33
(0.206, 0.245]    458
(0.245, 0.284]    659
(0.284, 0.323]     78
(0.323, 0.362]     10
Name: under18pct, dtype: int64

In [84]:
pd.qcut(popDF_wide['gte18pct'], 5).value_counts().sort_index()   

(0.637, 0.735]    248
(0.735, 0.745]    247
(0.745, 0.754]    248
(0.754, 0.765]    247
(0.765, 0.833]    248
Name: gte18pct, dtype: int64

There is a parameter called `labels`, which when set to 'False' will allow us to use a dummy int index for the bins instead of default labels (which uses the range of the bin as a text). Alternatively you can specify a list of names using the `labels` parameter as well.

In [85]:
popDF_wide['q_gte18pct'] = pd.qcut(popDF_wide['gte18pct'], 5, labels = False) 
display(popDF_wide)

abbrev  year  totalpop  under18pop      area   density  log_density  \
0        AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926   
1        AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834   
2        AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831   
3        AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823   
4        AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381   
...     ...   ...       ...         ...       ...       ...          ...   
1233     WY  2009  559851.0    134960.0   97818.0  5.723394     1.744562   
1234     WY  2010  564222.0    135351.0   97818.0  5.768079     1.752339   
1235     WY  2011  567329.0    135407.0   97818.0  5.799843     1.757831   
1236     WY  2012  576626.0    136526.0   97818.0  5.894886     1.774085   
1237     WY  2013  582658.0    137679.0   97818.0  5.956552     1.784492   

      gte18pop  avgstatepopbyyear  under18pct  gte18pct  gte18_pctgr  \
0     375788.0       4.894565e+06    0.320812  0.679188          NaN   
1     388013.0       4.960411e+06    0.319506  0.680494     0.032532   
2     403858.0       5.029691e+06    0.314025  0.685975     0.040836   
3     412244.0       5.096443e+06    0.312278  0.687722     0.020765   
4     415869.0       5.159330e+06    0.310685  0.689315     0.008793   
...        ...                ...         ...       ...          ...   
1233  424891.0       5.971383e+06    0.241064  0.758936     0.024990   
1234  428871.0       6.020144e+06    0.239890  0.760110     0.009367   
1235  431922.0       6.062868e+06    0.238675  0.761325     0.007114   
1236  440100.0       6.106254e+06    0.236767  0.763233     0.018934   
1237  444979.0       6.148922e+06    0.236295  0.763705     0.011086   

      q_gte18pct  
0              0  
1              0  
2              0  
3              0  
4              0  
...          ...  
1233           3  
1234           3  
1235           3  
1236           3  
1237           3  

[1238 rows x 13 columns]

## 2.14 Dummy coding

We can obtain dummy variables for the values in a column by calling the `pd.get_dummies()` function.\
You can also select a prefix for the dummy variables using the parameter `prefix`


In [86]:
dummy_gte18 = pd.get_dummies(popDF_wide['q_gte18pct'], prefix='G')
dummy_gte18

G_0  G_1  G_2  G_3  G_4
0       1    0    0    0    0
1       1    0    0    0    0
2       1    0    0    0    0
3       1    0    0    0    0
4       1    0    0    0    0
...   ...  ...  ...  ...  ...
1233    0    0    0    1    0
1234    0    0    0    1    0
1235    0    0    0    1    0
1236    0    0    0    1    0
1237    0    0    0    1    0

[1238 rows x 5 columns]

What if you want to add these dummies to your dataset?

You can use `DF.merge()`. merge has parameters `left_index` and `right_index` which can be be set to True if the columns being matched are the respective index


In [87]:
popDF_wide = popDF_wide.merge(dummy_gte18, how ='inner', left_index =True, right_index=True)
display(popDF_wide)

abbrev  year  totalpop  under18pop      area   density  log_density  \
0        AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926   
1        AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834   
2        AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831   
3        AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823   
4        AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381   
...     ...   ...       ...         ...       ...       ...          ...   
1233     WY  2009  559851.0    134960.0   97818.0  5.723394     1.744562   
1234     WY  2010  564222.0    135351.0   97818.0  5.768079     1.752339   
1235     WY  2011  567329.0    135407.0   97818.0  5.799843     1.757831   
1236     WY  2012  576626.0    136526.0   97818.0  5.894886     1.774085   
1237     WY  2013  582658.0    137679.0   97818.0  5.956552     1.784492   

      gte18pop  avgstatepopbyyear  under18pct  gte18pct  gte18_pctgr  \
0     375788.0       4.894565e+06    0.320812  0.679188          NaN   
1     388013.0       4.960411e+06    0.319506  0.680494     0.032532   
2     403858.0       5.029691e+06    0.314025  0.685975     0.040836   
3     412244.0       5.096443e+06    0.312278  0.687722     0.020765   
4     415869.0       5.159330e+06    0.310685  0.689315     0.008793   
...        ...                ...         ...       ...          ...   
1233  424891.0       5.971383e+06    0.241064  0.758936     0.024990   
1234  428871.0       6.020144e+06    0.239890  0.760110     0.009367   
1235  431922.0       6.062868e+06    0.238675  0.761325     0.007114   
1236  440100.0       6.106254e+06    0.236767  0.763233     0.018934   
1237  444979.0       6.148922e+06    0.236295  0.763705     0.011086   

      q_gte18pct  G_0  G_1  G_2  G_3  G_4  
0              0    1    0    0    0    0  
1              0    1    0    0    0    0  
2              0    1    0    0    0    0  
3              0    1    0    0    0    0  
4              0    1    0    0    0    0  
...          ...  ...  ...  ...  ...  ...  
1233           3    0    0    0    1    0  
1234           3    0    0    0    1    0  
1235           3    0    0    0    1    0  
1236           3    0    0    0    1    0  
1237           3    0    0    0    1    0  

[1238 rows x 18 columns]

## 2.15 Save DataFrame to a file

DataFrames can be saved to a number of formats - csv, html, json etc.

In [88]:
print(popDF_wide.count())
popDF_wide.dtypes

abbrev               1238
year                 1238
totalpop             1238
under18pop           1238
area                 1238
density              1238
log_density          1238
gte18pop             1238
avgstatepopbyyear    1238
under18pct           1238
gte18pct             1238
gte18_pctgr          1186
q_gte18pct           1238
G_0                  1238
G_1                  1238
G_2                  1238
G_3                  1238
G_4                  1238
dtype: int64


abbrev                object
year                   int64
totalpop             float64
under18pop           float64
area                 float64
density              float64
log_density          float64
gte18pop             float64
avgstatepopbyyear    float64
under18pct           float64
gte18pct             float64
gte18_pctgr          float64
q_gte18pct             int64
G_0                    uint8
G_1                    uint8
G_2                    uint8
G_3                    uint8
G_4                    uint8
dtype: object

In [89]:
popDF_wide.to_csv('popDF_wide.csv', header =True, index=False)

To read from a file into a DataFrame, there are methods such as `pd.read_csv()`, `pd.read_html()`, `pd.read_json()`.

When reading in a file, there are many options you can specify. `index_col` parameter is set to the column index that contains the index

In [90]:
pd.read_csv('popDF_wide.csv')


abbrev  year  totalpop  under18pop      area   density  log_density  \
0        AK  1990  553290.0    177502.0  656425.0  0.842884    -0.170926   
1        AK  1991  570193.0    182180.0  656425.0  0.868634    -0.140834   
2        AK  1992  588736.0    184878.0  656425.0  0.896882    -0.108831   
3        AK  1993  599434.0    187190.0  656425.0  0.913180    -0.090823   
4        AK  1994  603308.0    187439.0  656425.0  0.919081    -0.084381   
...     ...   ...       ...         ...       ...       ...          ...   
1233     WY  2009  559851.0    134960.0   97818.0  5.723394     1.744562   
1234     WY  2010  564222.0    135351.0   97818.0  5.768079     1.752339   
1235     WY  2011  567329.0    135407.0   97818.0  5.799843     1.757831   
1236     WY  2012  576626.0    136526.0   97818.0  5.894886     1.774085   
1237     WY  2013  582658.0    137679.0   97818.0  5.956552     1.784492   

      gte18pop  avgstatepopbyyear  under18pct  gte18pct  gte18_pctgr  \
0     375788.0       4.894565e+06    0.320812  0.679188          NaN   
1     388013.0       4.960411e+06    0.319506  0.680494     0.032532   
2     403858.0       5.029691e+06    0.314025  0.685975     0.040836   
3     412244.0       5.096443e+06    0.312278  0.687722     0.020765   
4     415869.0       5.159330e+06    0.310685  0.689315     0.008793   
...        ...                ...         ...       ...          ...   
1233  424891.0       5.971383e+06    0.241064  0.758936     0.024990   
1234  428871.0       6.020144e+06    0.239890  0.760110     0.009367   
1235  431922.0       6.062868e+06    0.238675  0.761325     0.007114   
1236  440100.0       6.106254e+06    0.236767  0.763233     0.018934   
1237  444979.0       6.148922e+06    0.236295  0.763705     0.011086   

      q_gte18pct  G_0  G_1  G_2  G_3  G_4  
0              0    1    0    0    0    0  
1              0    1    0    0    0    0  
2              0    1    0    0    0    0  
3              0    1    0    0    0    0  
4              0    1    0    0    0    0  
...          ...  ...  ...  ...  ...  ...  
1233           3    0    0    0    1    0  
1234           3    0    0    0    1    0  
1235           3    0    0    0    1    0  
1236           3    0    0    0    1    0  
1237           3    0    0    0    1    0  

[1238 rows x 18 columns]

# THE END